### PR TITLE
Implement `eth_getPendingTransactions`

### LINKED ISSUE #244: Implement `eth_getPendingTransactions`
__Proposal:__ Implement `eth_getPendingTransactions`

__Current behavior:__ empty implementation

__Desired behavior:__ implemented

__Use case:__ fetch transactions from mempool

### PR DESCRIPTION
**Pull request by yihuang**
Closes #244

## Description

- also refactored repeatitive tx code.

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/namespaces/eth/api.go [hunk 1]
```diff
@@ -864,8 +864,45 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
 func (e *PublicAPI) PendingTransactions() ([]*rpctypes.RPCTransaction, error) {
 	e.logger.Debugln("eth_getPendingTransactions")
 
-	// FIXME https://github.com/tharsis/ethermint/issues/244
-	return []*rpctypes.RPCTransaction{}, nil
+	var result []*rpctypes.RPCTransaction
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:867]:
```suggestion
	// nolint: prealloc
	var result []*rpctypes.RPCTransaction
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 2]
```diff
@@ -864,8 +864,45 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
...
+	txs, err := e.backend.PendingTransactions()
+	if err != nil {
+		return nil, err
+	}
+	for _, tx := range txs {
+		if len((*tx).GetMsgs()) != 1 {
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:873]: check if tx is nil

File: ethereum/rpc/namespaces/eth/api.go [hunk 3]
```diff
@@ -864,8 +864,45 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
...
+			e.logger.Debugln("invalid tx")
+			return nil, fmt.Errorf("invalid tx type: %T", tx)
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:875]: this should `continue` (unless all txs are `MsgEthereumTx`)

File: ethereum/rpc/namespaces/eth/api.go [hunk 4]
```diff
@@ -864,8 +864,45 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
...
+		}
+		msg, ok := (*tx).GetMsgs()[0].(*evmtypes.MsgEthereumTx)
+		if !ok {
+			e.logger.Debugln("invalid tx")
+			return nil, fmt.Errorf("invalid tx type: %T", tx)
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:880]: ditto

File: ethereum/rpc/namespaces/eth/api.go (hunk 1 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:867]: even better, I used prealloc now.

File: ethereum/rpc/namespaces/eth/api.go (hunk 2 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:873]: fixed

File: ethereum/rpc/namespaces/eth/api.go (hunk 3 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:875]: fixed

File: ethereum/rpc/namespaces/eth/api.go (hunk 4 above)
Comment (yihuang) [reply ethereum/rpc/namespaces/eth/api.go:880]: fixed

File: x/evm/types/msg.go [hunk 5]
```diff
@@ -37,6 +37,24 @@ func NewTx(
 	gasLimit uint64, gasPrice *big.Int, input []byte, accesses *ethtypes.AccessList,
 ) *MsgEthereumTx {
 	return newMsgEthereumTx(chainID, nonce, to, amount, gasLimit, gasPrice, input, accesses)
+
+}
+
+// UnwrapEthereumMsg extract MsgEthereumTx from wrapping sdk.Tx
+func UnwrapEthereumMsg(tx *sdk.Tx) (*MsgEthereumTx, error) {
```
Comment (fedekunze) [inline x/evm/types/msg.go:44]: can we add a test for this function? 🙏 and maybe move it to `utils.go`

File: ethereum/rpc/namespaces/eth/api.go [hunk 6]
```diff
@@ -719,31 +682,18 @@ func (e *PublicAPI) GetTransactionByBlockHashAndIndex(hash common.Hash, idx hexu
...
-	}
-
-	txHash := msg.AsTransaction().Hash()
-
-	txData, err := evmtypes.UnpackTxData(msg.Data)
+	msg, err := evmtypes.UnwrapEthereumMsg(&tx)
 	if err != nil {
-		e.logger.WithError(err).Debugln("decoding failed")
-		return nil, fmt.Errorf("failed to unpack tx data: %w", err)
+		e.logger.Debugln("invalid tx")
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:687]:
```suggestion
		e.logger.WithError(err).Debugln("invalid tx")
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 7]
```diff
@@ -666,33 +644,18 @@ func (e *PublicAPI) GetTransactionByHash(hash common.Hash) (*rpctypes.RPCTransac
...
-	msg, ok := tx.GetMsgs()[0].(*evmtypes.MsgEthereumTx)
-	if !ok {
-		e.logger.Debugln("invalid tx")
-		return nil, fmt.Errorf("invalid tx type: %T", tx)
-	}
-
-	from, err := msg.GetSender(e.chainIDEpoch)
+	msg, err := evmtypes.UnwrapEthereumMsg(&tx)
 	if err != nil {
+		e.logger.Debugln("invalid tx")
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:649]:
```suggestion
		e.logger.WithError(err).Debugln("invalid tx")
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 8]
```diff
@@ -770,31 +720,18 @@ func (e *PublicAPI) GetTransactionByBlockNumberAndIndex(blockNum rpctypes.BlockN
...
-	}
-
-	txHash := msg.AsTransaction().Hash()
-
-	txData, err := evmtypes.UnpackTxData(msg.Data)
+	msg, err := evmtypes.UnwrapEthereumMsg(&tx)
 	if err != nil {
-		e.logger.WithError(err).Debugln("decoding failed")
-		return nil, fmt.Errorf("failed to unpack tx data: %w", err)
+		e.logger.Debugln("invalid tx")
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:725]:
```suggestion
		e.logger.WithError(err).Debugln("invalid tx")
```

File: ethereum/rpc/namespaces/eth/api.go [hunk 9]
```diff
@@ -820,15 +757,10 @@ func (e *PublicAPI) GetTransactionReceipt(hash common.Hash) (map[string]interfac
...
-	if len(tx.GetMsgs()) != 1 {
-		e.logger.Debugln("invalid tx")
-		return nil, fmt.Errorf("invalid tx type: %T", tx)
-	}
-
-	msg, ok := tx.GetMsgs()[0].(*evmtypes.MsgEthereumTx)
-	if !ok {
+	msg, err := evmtypes.UnwrapEthereumMsg(&tx)
+	if err != nil {
 		e.logger.Debugln("invalid tx")
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/eth/api.go:762]:
```suggestion
		e.logger.WithError(err).Debugln("invalid tx")
```

File: x/evm/types/msg.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/msg.go]: done

**Review by fedekunze — APPROVED** [review]

File: x/evm/types/utils_test.go [hunk 10]
```diff
@@ -40,15 +46,36 @@ func TestEvmDataEncoding(t *testing.T) {
...
+	_, err = evmtypes.UnwrapEthereumMsg(&tx)
+	require.NotNil(t, err)
+
+	msg := evmtypes.NewTx(big.NewInt(1), 0, &common.Address{}, big.NewInt(0), 0, big.NewInt(0), []byte{}, nil)
+	err = builder.SetMsgs(msg)
+
+	tx = builder.GetTx().(sdk.Tx)
+	msg_, err := evmtypes.UnwrapEthereumMsg(&tx)
+	require.Nil(t, err)
+	require.Equal(t, *msg_, msg)
```
Comment (fedekunze) [inline x/evm/types/utils_test.go:80]: this seems to be failing

**Pull request merged**